3.7/ubuntu/Dockerfile

```diff
@@ -0,0 +1,138 @@
+    tar -zxvf $OPENSSL_PATH.tar.gz --directory $OPENSSL_PATH --strip-components=1
+# Configure OpenSSL for compilation
+RUN cd $OPENSSL_PATH && \
+    ./config --prefix=$OPENSSL_INSTALL_DIR --openssldir=$OPENSSL_CONFIG_DIR
+# Compile & install OpenSSL
+RUN cd $OPENSSL_PATH && make -j $(getconf _NPROCESSORS_ONLN) && make test install
+# Record OpenSSL artefacts size
+RUN du -skh $OPENSSL_INSTALL_DIR $OPENSSL_CONFIG_DIR >> /BUILD_ARTEFACTS
+
+# Download, verify & extract OTP_SOURCE
+# Fail fast if SOURCE URL returns HTTP errors
+RUN curl --verbose --head --fail --fail-early $OTP_SOURCE_URL 1>/dev/null && \
+    curl --verbose --location --silent --fail --fail-early --output $OTP_PATH.tar.gz $OTP_SOURCE_URL && \
+    echo "$OTP_SOURCE_SHA256  $OTP_PATH.tar.gz" > $OTP_PATH.sha256 && \
+    sha256sum --check --strict $OTP_PATH.sha256 && \
+    tar -zxvf $OTP_PATH.tar.gz --directory $OTP_PATH --strip-components=1
+
+# Required for Erlang/OTP makefiles to find the absolute path for the installation
+ENV ERL_TOP=$OTP_PATH
+
+# Configure Erlang/OTP for compilation, disable unused features & applications
+# http://erlang.org/doc/applications.html
+RUN cd $OTP_PATH && \
+    ./otp_build autoconf && \
+    ./configure \
+      --prefix=$OTP_INSTALL_DIR \
+      --disable-sctp \
+      --enable-hipe \
+      --enable-lock-counter \
+      --with-microstate-accounting=extra \
+      --disable-dynamic-ssl-lib \
+      --with-ssl=$OPENSSL_INSTALL_DIR \
+      --without-odbc \
+      --without-diameter \
+      --without-erl_interface \
+      --without-ftp \
+      --without-jinterface \
+      --without-megaco \
+      --without-ssh \
+      --without-tftp \
+      --without-wx \
+      --without-debugger \
+      --without-dialyzer \
+      --without-et \
+      --without-observer \
+      --without-reltool \
+      --without-common_test \
+      --without-eunit \
+      --without-edoc \
+      --without-erl_docgen
+
+# Compile & install Erlang/OTP
+RUN cd $OTP_PATH && make -j $(getconf _NPROCESSORS_ONLN) && make install
+# Record Erlang/OTP artefacts size
+RUN du -skh $OTP_INSTALL_DIR >> /BUILD_ARTEFACTS
+
+# Check that Erlang/OTP crypto & ssl were compiled against OpenSSL correctly
+RUN $OTP_INSTALL_DIR/bin/erl -noshell -eval 'io:format("~p~n~n~p~n~n", [crypto:supports(), ssl:versions()]), init:stop().'
+
+# Start with the unmodified release image
+FROM release
+# Only copy the build artefacts required by RabbitMQ
+COPY --from=openssl-otp /opt /opt
+COPY --from=openssl-otp /usr/local/ssl /usr/local/ssl
+# Add OpenSSL to PATH
+ENV PATH=/opt/openssl/bin:$PATH
+ENV LD_LIBRARY_PATH=/opt/openssl/lib
+# Ensure OpenSSL works by listing all commands
+RUN openssl help
+# Add Erlang/OTP to PATH
+ENV PATH=/opt/otp/bin:$PATH
+# Ensure Erlang/OTP works by listing all loaded modules
+RUN erl -noshell -eval 'io:format("~p~n", [code:all_loaded()]), init:stop().'
+
+# TODO
+# Install RabbitMQ
+# Ensure RabbitMQ starts & rabbitmqctl works correctly
```

Comment:
Here's the modifications I was playing with (based on our existing image, so take it for what it's worth -- all I've changed from what I was playing with was to copy your list of Erlang compilations flags from above):

```dockerfile
FROM debian:stretch-slim

RUN set -eux; \
	apt-get update; \
	apt-get install -y --no-install-recommends \
		gnupg \
		dirmngr \
	; \
	rm -rf /var/lib/apt/lists/*

# add our user and group first to make sure their IDs get assigned consistently, regardless of whatever dependencies get added
RUN groupadd -r rabbitmq && useradd -r -d /var/lib/rabbitmq -m -g rabbitmq rabbitmq

# grab gosu for easy step-down from root
ENV GOSU_VERSION 1.10
RUN set -eux; \
	\
	fetchDeps=' \
		ca-certificates \
		wget \
	'; \
	apt-get update; \
	apt-get install -y --no-install-recommends $fetchDeps; \
	rm -rf /var/lib/apt/lists/*; \
	\
	dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"; \
	wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch"; \
	wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc"; \
	\
# verify the signature
	export GNUPGHOME="$(mktemp -d)"; \
	gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4; \
	gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu; \
	command -v gpgconf && gpgconf --kill all || :; \
	rm -rf "$GNUPGHOME" /usr/local/bin/gosu.asc; \
	\
	chmod +x /usr/local/bin/gosu; \
# verify that the binary works
	gosu nobody true; \
	\
	apt-get purge -y --auto-remove $fetchDeps

# RabbitMQ 3.6.15+ requires Erlang 19.3+, 3.7.x needs 20.3+, 3.8.x needs 21.2+; https://www.rabbitmq.com/which-erlang.html
# https://github.com/docker-library/rabbitmq/issues/277
ENV ERLANG_VERSION 21.2
# https://github.com/rabbitmq/erlang-debian-package/blob/HEAD/debian/changelog#L1
# https://github.com/erlang/otp/releases
ENV ERLANG_SHA256 5d2cb28232a60ce88c6478fcf5d6aa5be353555e02f3cf96ed93c9bae7522448

RUN set -eux; \
	\
	savedAptMark="$(apt-mark showmanual)"; \
	apt-get update; \
	\
	apt-get install -y --no-install-recommends ca-certificates wget; \
	wget -O erlang.tgz "https://github.com/erlang/otp/archive/OTP-${ERLANG_VERSION}.tar.gz" --progress=dot:giga; \
	echo "$ERLANG_SHA256 *erlang.tgz" | sha256sum -c -; \
	mkdir -p /usr/src/erlang; \
	tar -xf erlang.tgz -C /usr/src/erlang --strip-components=1; \
	rm erlang.tgz; \
	\
	apt-get install -y --no-install-recommends \
# https://github.com/rabbitmq/erlang-debian-package/blob/HEAD/debian/control
		autoconf \
		dpkg-dev \
		gcc \
		libc6-dev \
		libncurses5-dev \
		libssl-dev \
	; \
# https://github.com/rabbitmq/erlang-debian-package/blob/HEAD/debian/rules
	cd /usr/src/erlang; \
	./otp_build autoconf; \
	DEB_HOST_GNU_TYPE="$(dpkg-architecture --query DEB_HOST_GNU_TYPE)"; \
	DEB_BUILD_GNU_TYPE="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)"; \
	export CFLAGS="$(dpkg-buildflags --get CFLAGS)"; \
	./configure \
		--host="$DEB_HOST_GNU_TYPE" \
		--build="$DEB_BUILD_GNU_TYPE" \
		--disable-dynamic-ssl-lib \
		--disable-sctp \
		--disable-silent-rules \
		--enable-clock-gettime \
		--enable-dynamic-ssl-lib \
		--enable-ethread-pre-pentium4-compatibility \
		--enable-hipe \
		--enable-hybrid-heap \
		--enable-kernel-poll \
		--enable-lock-counter \
		--enable-shared-zlib \
		--enable-threads --enable-smp-support \
		--with-microstate-accounting=extra \
		--without-common_test \
		--without-debugger \
		--without-dialyzer \
		--without-diameter \
		--without-edoc \
		--without-erl_docgen \
		--without-erl_interface \
		--without-et \
		--without-eunit \
		--without-ftp \
		--without-jinterface \
		--without-megaco \
		--without-observer \
		--without-odbc \
		--without-reltool \
		--without-ssh \
		--without-ssl-zlib \
		--without-tftp \
		--without-wx \
		--with-ssl-rpath=no \
	; \
	nproc="$(nproc)"; \
	make -j "$nproc" GEN_OPT_FLGS='-O2 -fno-strict-aliasing'; \
	make -j "$nproc" install; \
	\
	cd /; rm -rf /usr/src/erlang; \
	\
	apt-mark auto '.*' > /dev/null; \
	apt-mark manual $savedAptMark; \
	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false; \
	rm -rf /var/lib/apt/lists/*; \
	\
	erl -version
RUN false # forcing the build to fail here so I can mess with the Erlang layer more actively

# get logs to stdout (thanks @dumbbell for pushing this upstream! :D)
ENV RABBITMQ_LOGS=- RABBITMQ_SASL_LOGS=-
# https://github.com/rabbitmq/rabbitmq-server/commit/53af45bf9a162dec849407d114041aad3d84feaf

# /usr/sbin/rabbitmq-server has some irritating behavior, and only exists to "su - rabbitmq /usr/lib/rabbitmq/bin/rabbitmq-server ..."
ENV PATH /usr/lib/rabbitmq/bin:$PATH

# gpg: key 6026DFCA: public key "RabbitMQ Release Signing Key <info@rabbitmq.com>" imported
ENV RABBITMQ_GPG_KEY 0A9AF2115F4687BD29803A206B73A36E6026DFCA

ENV RABBITMQ_VERSION 3.7.8
ENV RABBITMQ_GITHUB_TAG v3.7.8
ENV RABBITMQ_DEBIAN_VERSION 3.7.8-1

RUN set -eux; \
	\
	apt-get update; \
	apt-get install -y --no-install-recommends ca-certificates wget; \
	\
	wget -O rabbitmq-server.deb.asc "https://github.com/rabbitmq/rabbitmq-server/releases/download/$RABBITMQ_GITHUB_TAG/rabbitmq-server_${RABBITMQ_DEBIAN_VERSION}_all.deb.asc"; \
	wget -O rabbitmq-server.deb     "https://github.com/rabbitmq/rabbitmq-server/releases/download/$RABBITMQ_GITHUB_TAG/rabbitmq-server_${RABBITMQ_DEBIAN_VERSION}_all.deb"; \
	\
	apt-get purge -y --auto-remove ca-certificates wget; \
	\
	export GNUPGHOME="$(mktemp -d)"; \
	gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys "$RABBITMQ_GPG_KEY"; \
	gpg --batch --verify rabbitmq-server.deb.asc rabbitmq-server.deb; \
	command -v gpgconf && gpgconf --kill all || :; \
	rm -rf "$GNUPGHOME"; \
	\
	apt install -y --no-install-recommends ./rabbitmq-server.deb; \
	dpkg -l | grep rabbitmq-server; \
	rm -f rabbitmq-server.deb*; \
	\
	rm -rf /var/lib/apt/lists/*

# warning: the VM is running with native name encoding of latin1 which may cause Elixir to malfunction as it expects utf8. Please ensure your locale is set to UTF-8 (which can be verified by running "locale" in your shell)
ENV LANG C.UTF-8

# set home so that any `--user` knows where to put the erlang cookie
ENV HOME /var/lib/rabbitmq

RUN mkdir -p /var/lib/rabbitmq /etc/rabbitmq /var/log/rabbitmq /tmp/rabbitmq-ssl \
	&& chown -R rabbitmq:rabbitmq /var/lib/rabbitmq /etc/rabbitmq /var/log/rabbitmq /tmp/rabbitmq-ssl \
	&& chmod -R 777 /var/lib/rabbitmq /etc/rabbitmq /var/log/rabbitmq /tmp/rabbitmq-ssl
VOLUME /var/lib/rabbitmq

# add a symlink to the .erlang.cookie in /root so we can "docker exec rabbitmqctl ..." without gosu
RUN ln -sf /var/lib/rabbitmq/.erlang.cookie /root/

RUN ln -sf "/usr/lib/rabbitmq/lib/rabbitmq_server-$RABBITMQ_VERSION/plugins" /plugins

COPY docker-entrypoint.sh /usr/local/bin/
RUN ln -s usr/local/bin/docker-entrypoint.sh / # backwards compat
ENTRYPOINT ["docker-entrypoint.sh"]

EXPOSE 4369 5671 5672 25672
CMD ["rabbitmq-server"]
```

One thing I'm really concerned about is that even with your more-aggressive list of flags to Erlang's `./configure`, that "install Erlang" line is still almost the size of our entire previous image (~122MB for that layer vs ~125MB for the full current `rabbitmq:3.7` image). :cold_sweat:

Reply:
This looks good! I am wondering whether we should prioritize this change in the Debian image over the final Ubuntu image that is still WIP. What do you think?

You will notice the `RUN du` commands in the initial Dockerfile which capture the install sizes of both Erlang/OTP & OpenSSL. In the initial comment, I have this:

>  I suspect that the final Erlang/OTP release can be shrunk further (it currently stands 130MB), but this is a minor concern right now.

I didn't think this will be a major concern. I will check if there are any quick learnings that I can steal from the code that is used for producing https://dl.bintray.com/rabbitmq/debian/dists/bionic

I have a `gosu`-related question: why not use `USER rabbitmq` in the Dockerfile and then just `CMD rabbitmq-server` ?

Reply:
FTR the Debian packages are now distributed to `https://bintray.com/rabbitmq-erlang`.

Note that Debian packages do not include the source, slice-and-dice OTP and do not enable things such as extra metrics in microstate accounting. All 3 strategies can be applied here, in fact, @gerhard has discovered that microstate accounting, even though very useful, significantly increases final OTP build size. So it likely will be dropped.

Reply:
> I didn't think this will be a major concern. I will check if there are any quick learnings that I can steal from the code that is used for producing https://dl.bintray.com/rabbitmq/debian/dists/bionic

I think the difference there is that when we use Erlang Debian packages, we only install a minimum set of packages (even though the Debian package source built everything), so the approach I was taking previously was to remove some things post-build based on what I didn't think was used, but there's probably a better and/or more methodical approach we can take there.

> I have a `gosu`-related question: why not use `USER rabbitmq` in the Dockerfile and then just `CMD rabbitmq-server`?

That'd be so that if users don't specify any `--user` and bind-mount their own storage, we can fix permissions:

https://github.com/docker-library/rabbitmq/blob/3740043813fc8a0f0b58451c94e181f441b37dc5/3.7/docker-entrypoint.sh#L83-L112

If the image is run as non-root, it will simply skip those lines, which allows for users to control their own destiny (but does require that they make sure their volumes/mounts already have appropriate permissions for the UID/GID they specify).

Reply:
It would make sense to use [rabbitmq/erlang-debian-package](https://github.com/rabbitmq/erlang-debian-package) to compile Erlang/OTP for this Docker image, but I'm leaning towards doing this as a future refinement. The main reason is that I am not familiar with that codebase and would prefer to work with @dumbbell on this. After all, we don't need deb packages, we simply need to install the Erlang/OTP artefacts required by RabbitMQ.

As for the final image size, this is what I've found in regards to Erlang/OTP:

* installing all Erlang/OTP applications that rabbitmq-server deb package depends on uses `40MB`
* with the initial Erlang/OTP compilation approach, we end up with `130M /opt/otp/`
* dropping lock counting saves 45MB: `45M /opt/otp/lib/erlang/erts-10.2.1/bin/beam.lcnt.smp`
* deleting all Erlang/OTP application source code saves 19MB:
```
find /opt/otp/lib/erlang/lib -maxdepth 2 -mindepth 2 -name "src" -exec du -ch {} + | grep total$
19M total
```
* the above 2 steps result in `66MB` for Erlang/OTP vs the `40MB` with Debian packages
* pretty sure we could reduce the `66MB` further, this is what takes up the majority of the `66MB`:
```
64M	total
27M	/opt/otp/lib/erlang/erts-10.2.1/bin
6.8M	/opt/otp/lib/erlang/lib/crypto-4.4
6.5M	/opt/otp/lib/erlang/lib/hipe-3.18.2
4.0M	/opt/otp/lib/erlang/lib/stdlib-3.7
3.2M	/opt/otp/lib/erlang/lib/snmp-5.2.12
1.9M	/opt/otp/lib/erlang/lib/tools-3.0.2
1.9M	/opt/otp/lib/erlang/lib/kernel-6.2
1.8M	/opt/otp/lib/erlang/lib/public_key-1.6.4
1.7M	/opt/otp/lib/erlang/lib/inets-7.0.3
1.7M	/opt/otp/lib/erlang/lib/compiler-7.3
1.6M	/opt/otp/lib/erlang/lib/xmerl-1.3.18
1.3M	/opt/otp/lib/erlang/lib/mnesia-4.15.5
1.2M	/opt/otp/lib/erlang/lib/ssl-9.1.1
1.2M	/opt/otp/lib/erlang/lib/asn1-5.0.8
544K	/opt/otp/lib/erlang/lib/syntax_tools-2.1.6
476K	/opt/otp/lib/erlang/lib/runtime_tools-1.13.1
448K	/opt/otp/lib/erlang/lib/erts-10.2.1
408K	/opt/otp/lib/erlang/lib/sasl-3.3
276K	/opt/otp/lib/erlang/lib/os_mon-2.4.7
256K	/opt/otp/lib/erlang/lib/parsetools-2.1.8
180K	/opt/otp/lib/erlang/lib/eldap-1.2.6
120K	/opt/otp/lib/erlang/lib/otp_mibs-1.2.1
```
* most space is taken by `25M /opt/otp/lib/erlang/erts-10.2.1/bin/beam.smp`, while the Debian one is `3.1M /usr/lib/erlang/erts-9.2/bin/beam.smp` - I don't know yet why this is the case

@tianon I get the `gosu` bit now, thank you.